File: epi_judge_python/buy_and_sell_stock.py
from typing import List

from test_framework import generic_test


def buy_and_sell_stock_once(prices: List[float]) -> float:
    minv = prices[0]
    diff = 0
    for i, p in enumerate(prices):
        diff = max(p - minv, diff)
        minv = min(p, minv)
    return diff


if __name__ == "__main__":
    exit(
        generic_test.generic_test_main(
            "buy_and_sell_stock.py", "buy_and_sell_stock.tsv", buy_and_sell_stock_once
        )
    )

File: epi_judge_cpp/buy_and_sell_stock.cc
#include <algorithm>
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;
double BuyAndSellStockOnce(const vector<double> &prices) {
  // TODO - you fill in here.
  // return 0.0;
  double profit{0.0}, sell{prices.back()};

  for (auto rit = prices.rbegin(); rit != prices.rend(); rit++) {
    if (*rit > sell) {
      sell = *rit;
      continue;
    }
    profit = std::max(profit, sell - *rit);
  }
  return profit;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"prices"};
  return GenericTestMain(args, "buy_and_sell_stock.cc",
                         "buy_and_sell_stock.tsv", &BuyAndSellStockOnce,
                         DefaultComparator{}, param_names);
}
